### ISSUE TITLE
bug in Utilities::MPI::Partitioner?

### ISSUE THREAD

**Issue by davydden** [dealii/dealii#7095]
I was looking at the code in partitioner and this 

https://github.com/dealii/dealii/blob/master/source/base/partitioner.cc#L214-L238 

part looks suspicious (perhaps I misunderstand something). After `MPI_Allgather`, starting from the second element `first_index` contains `local_range_data.second` from each process (order both by rank and value). I am not sure how checks like `first_index[i] == 0` could ever be hit after we found `first_proc_with_nonzero_dofs`.  Also looking at the way `first_index` is used in the code I don't immediately see a reason to treat the case of no locally-owned DoFs somehow specially. Checks like 

https://github.com/dealii/dealii/blob/master/source/base/partitioner.cc#L254-L255

should be ok even we have a couple of `end's` with the same value, meaning no locally owned DoFs on some processes.

@kronbichler ping.

**Comment by kronbichler** [comment]
> Also, should not `first_index[n_procs] == global_size` hold after Allgather?

No, it does not. If the last processor has no indices in the locally owned range, the value of `local_range_data` on that processor will be `[0, 0)` (an invalid range, the default constructor) and thus the last index will be wrong. An empty `IndexSet` on a certain processor does not have any notion where in the global vector partitioning its range fits in. However, some of the code further down assumes that it is possible to decide where the locally owned range would have been if it contained entries. What this code does is to ensure that the start point is always the end index of the processor immediately before. So the fixup is not for the special case `start == end` in particular, but the reason that the initial value is `[0,0)`. There might be another solution than line 214-238 to fix up the case of empty locally owned ranges, but back in 2010/2011 when we wrote it we hit assertions elsewhere without having the indices sorted and logically correct.

Is there anything particular that you are unhappy about with this code so we should look for another solution?

One thing I believe could be simplified is the fact that we could write the section 
https://github.com/dealii/dealii/blob/master/source/base/partitioner.cc#L218-L228
simpler as 
```
          for (unsigned int i = 0; i < n_procs; ++i)
            if (first_index[i] == 0)
              first_index[i] = first_index[i - 1];
```
